File: apps/api/app/services/domain.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date, datetime, timezone
from decimal import Decimal
from typing import Any
from uuid import uuid4

from app.core.enums import (
    FundingStatus,
    ProjectStatus,
    RevisionStatus,
    RevisionType,
    SanctionStatus,
    TrancheStatus,
    TransactionType,
)
from app.core.money import ZERO, money


def new_id() -> str:
    return str(uuid4())


@dataclass
class ProjectParticipant:
    id: str = field(default_factory=new_id)
    project_id: str = ""
    role: str = "project_member"
    full_name: str = ""
    email: str | None = None
    phone: str | None = None
    department: str | None = None
    organization: str | None = None
    is_primary: bool = False
    start_date: date | None = None
    end_date: date | None = None
    notes: str | None = None


@dataclass
class Project:
    id: str = field(default_factory=new_id)
    project_code: str = ""
    title: str = ""
    original_reference: str | None = None
    short_title: str | None = None
    description: str | None = None
    institution: str | None = None
    school: str | None = None
    department: str | None = None
    academic_year: str | None = None
    cohort: str | None = None
    category: str | None = None
    domain: str | None = None
    technology_readiness_level: str | None = None
    prototype_status: str | None = None
    publication_status: str | None = None
    patent_status: str | None = None
    startup_status: str | None = None
    project_status: ProjectStatus = ProjectStatus.DRAFT
    funding_status: FundingStatus = FundingStatus.NOT_SANCTIONED
    start_date: date | None = None
    expected_completion_date: date | None = None
    actual_completion_date: date | None = None
    closure_notes: str | None = None
    remarks: str | None = None
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    created_by: str | None = None
    updated_by: str | None = None
    version: int = 1
    participants: list[ProjectParticipant] = field(default_factory=list)


@dataclass
class FundingSanction:
    id: str = field(default_factory=new_id)
    project_id: str = ""
    sanction_reference: str = ""
    sanction_date: date | None = None
    amount: Decimal = ZERO
    funding_source: str | None = None
    financial_year: str | None = None
    status: SanctionStatus = SanctionStatus.DRAFT
    approved_by: str | None = None
    approved_at: datetime | None = None
    remarks: str | None = None

    def __post_init__(self) -> None:
        self.amount = money(self.amount)


@dataclass
class FundingRevision:
    id: str = field(default_factory=new_id)
    project_id: str = ""
    revision_number: int = 1
    revision_type: RevisionType = RevisionType.INCREASE
    revision_date: date | None = None
    amount: Decimal = ZERO
    approval_reference: str | None = None
    reason: str | None = None
    status: RevisionStatus = RevisionStatus.DRAFT
    approved_by: str | None = None
    approved_at: datetime | None = None
    remarks: str | None = None

    def __post_init__(self) -> None:
        self.amount = money(self.amount)


@dataclass
class Tranche:
    id: str = field(default_factory=new_id)
    project_id: str = ""
    sequence_number: int = 1
    transaction_type: TransactionType = TransactionType.ADVANCE
    purchase_order_number: str | None = None
    purchase_order_received_date: date | None = None
    request_date: date | None = None
    requested_amount: Decimal = ZERO
    approved_amount: Decimal = ZERO
    approval_date: date | None = None
    expected_disbursement_date: date | None = None
    actual_disbursement_date: date | None = None
    disbursed_amount: Decimal = ZERO
    refund_amount: Decimal = ZERO
    utilized_amount: Decimal = ZERO
    payment_mode: str | None = None
    payment_reference: str | None = None
    bill_status: str | None = None
    utilization_certificate_status: str | None = None
    status: TrancheStatus = TrancheStatus.DRAFT
    remarks: str | None = None

    def __post_init__(self) -> None:
        self.requested_amount = money(self.requested_amount)
        self.approved_amount = money(self.approved_amount)
        self.disbursed_amount = money(self.disbursed_amount)
        self.refund_amount = money(self.refund_amount)
        self.utilized_amount = money(self.utilized_amount)


@dataclass
class AuditEvent:
    id: str = field(default_factory=new_id)
    entity_type: str = ""
    entity_id: str = ""
    action: str = ""
    actor_id: str | None = None
    timestamp: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    previous_values: dict[str, Any] = field(default_factory=dict)
    new_values: dict[str, Any] = field(default_factory=dict)
    reason: str | None = None
    source: str = "system"
    request_id: str | None = None

File: apps/web/src/types/domain.ts
export type Role = "administrator" | "fund_administrator" | "fund_reviewer" | "auditor" | "viewer" | string;

export interface User {
  id: string;
  email: string;
  fullName: string;
  role: Role;
  roleLabel?: string;
  isActive?: boolean;
}

export interface FinancialSummary {
  initialSanctionedAmount?: string;
  approvedFundingIncreases?: string;
  approvedFundingReductions?: string;
  totalSanctionedAmount: string;
  totalRequestedAmount?: string;
  totalApprovedTrancheAmount?: string;
  grossDisbursedAmount?: string;
  totalRefundedAmount?: string;
  netDisbursedAmount: string;
  totalUtilizedAmount?: string;
  availableSanctionedBalance: string;
  unutilizedDisbursedBalance?: string;
  pendingApprovedAmount: string;
  trancheCount: number;
  reconciliationStatus: string;
}

export interface Participant {
  id?: string;
  role: string;
  roleLabel?: string;
  fullName: string;
  email?: string | null;
  phone?: string | null;
  department?: string | null;
  organization?: string | null;
  isPrimary?: boolean;
  startDate?: string | null;
  endDate?: string | null;
  notes?: string | null;
}

export interface ProjectRow {
  id: string;
  projectCode: string;
  project_code?: string;
  title: string;
  shortTitle?: string | null;
  description?: string | null;
  institution?: string | null;
  school?: string | null;
  department?: string | null;
  academicYear?: string | null;
  cohort?: string | null;
  category?: string | null;
  domain?: string | null;
  technologyReadinessLevel?: string | null;
  prototypeStatus?: string | null;
  publicationStatus?: string | null;
  patentStatus?: string | null;
  startupStatus?: string | null;
  status: string;
  projectStatus?: string;
  fundingStatus: string;
  startDate?: string | null;
  expectedCompletionDate?: string | null;
  actualCompletionDate?: string | null;
  closureNotes?: string | null;
  remarks?: string | null;
  participants?: Participant[];
  version: number;
  summary: FinancialSummary;
}

export interface ProjectDetail extends ProjectRow {
  sanctions: Record<string, unknown>[];
  fundingRevisions: Record<string, unknown>[];
  tranches: TrancheRow[];
}

export interface TrancheRow {
  id: string;
  projectId: string;
  projectCode?: string;
  projectTitle?: string;
  sequenceNumber: number;
  transactionType: string;
  transactionTypeLabel?: string;
  requestedAmount: string;
  approvedAmount: string;
  disbursedAmount: string;
  refundAmount: string;
  utilizedAmount: string;
  paymentReference?: string | null;
  paymentMode?: string | null;
  purchaseOrderNumber?: string | null;
  requestDate?: string | null;
  approvalDate?: string | null;
  expectedDisbursementDate?: string | null;
  actualDisbursementDate?: string | null;
  billStatus?: string | null;
  utilizationCertificateStatus?: string | null;
  status: string;
  statusLabel?: string;
  remarks?: string | null;
}

export interface ReconciliationIssue {
  id?: string;
  issueType: string;
  severity: "low" | "medium" | "high" | "critical" | string;
  projectId?: string;
  projectCode: string;
  projectTitle?: string;
  description: string;
  financialImpact?: string | null;
  status?: string;
  suggestedAction?: string;
}

export interface AuditEvent {
  id: string;
  entityType: string;
  entityId: string;
  action: string;
  actorId?: string | null;
  timestamp?: string | null;
  reason?: string | null;
  previousValues: Record<string, unknown>;
  newValues: Record<string, unknown>;
}

export interface SearchResult {
  type: string;
  label: string;
  description?: string | null;
  to: string;
  projectId?: string;
  trancheId?: string;
}

export interface ImportRowResult {
  id: string;
  rowNumber: number;
  status: string;
  proposedAction: string;
  duplicate: boolean;
  errors: string[];
  warnings: string[];
  rawValues: Record<string, unknown>;
  normalizedValues: Record<string, unknown>;
  result: Record<string, unknown>;
}

export interface ImportBatch {
  id: string;
  importType: string;
  filename: string;
  status: string;
  rowsDetected: number;
  validRows: number;
  invalidRows: number;
  duplicateRows: number;
  existingRecordsMatched: number;
  proposedCreates: number;
  proposedUpdates: number;
  committedRows: number;
  failedRows: number;
  skippedRows: number;
  rows: ImportRowResult[];
}

export interface AIProposal {
  id: string;
  action: string;
  status: string;
  targetEntityType?: string | null;
  targetEntityId?: string | null;
  currentValues: Record<string, unknown>;
  proposedValues: Record<string, unknown>;
  validationResult?: { valid?: boolean; warnings?: string[]; errors?: string[] };
  message?: string | null;
  expiresAt?: string | null;
  result?: Record<string, unknown>;
}

export interface AIResponse {
  kind: "answer" | "proposal" | "clarification" | "error" | "result" | "export" | string;
  message: string;
  proposal?: AIProposal;
  data?: Record<string, unknown> | Array<Record<string, unknown>>;
  download_url?: string;
}

export interface AIMessage {
  id: string;
  conversationId: string;
  role: "user" | "assistant" | string;
  content: string;
  responseKind?: string | null;
  action?: string | null;
  metadata?: AIResponse | Record<string, unknown>;
  proposalId?: string | null;
  createdAt?: string | null;
}

export interface AIConversation {
  id: string;
  title: string;
  projectId?: string | null;
  projectCode?: string | null;
  archived: boolean;
  messages?: AIMessage[];
  createdAt?: string | null;
  updatedAt?: string | null;
}

export interface AIConversationSendResult {
  conversation: AIConversation;
  response: AIResponse;
}

export interface SettingsPayload {
  profile: User;
  roles: Array<{ value: string; label: string }>;
  ai: {
    enabled: boolean;
    baseUrlConfigured: boolean;
    modelConfigured: boolean;
    model?: string | null;
    timeoutSeconds: number;
    maxTokens: number;
    temperature: number;
  };
  application: {
    name: string;
    version: string;
    license: string;
  };
}
